File: stack/reverse_polish_notation/rpn.js
/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function(tokens) {
  if (tokens.length === 1) return parseInt(tokens[0]);
  stack = [];
  tokens.forEach(c => {
      if (c == '+') stack.push(stack.pop() + stack.pop());
      else if (c == '*') stack.push(stack.pop() * stack.pop());
      else if (c == '/') {
          const factor = stack.pop();
          stack.push(Math.trunc(stack.pop() / factor));
      }
      else if (c == '-') {
          const factor = stack.pop();
          stack.push(stack.pop() - factor);
      }
      else stack.push(parseInt(c));
  });
  return stack.pop();
}

File: stack/reverse_polish_notation/rpn.py
import math

class Solution:
    def evalRPN(self, tokens) -> int:
        if not tokens:
            return 0
        if len(tokens) == 1:
            return int(tokens[0])
        stack = []
        for val in tokens:
            if val == '+':
                stack.append(stack.pop() + stack.pop())
            elif val == '-':
                factor = stack.pop()
                stack.append(stack.pop() - factor)
            elif val == '/':
                factor = stack.pop()
                stack.append(math.trunc(stack.pop() / factor))
            elif val == '*':
                stack.append(stack.pop() * stack.pop())
            else:
                stack.append(int(val))
        return stack.pop()


